### ISSUE TITLE
BytesMut memory usage grows without bound

### ISSUE THREAD

**Issue by nzentzis** [carllerche/bytes#193]
I'm working on a high-throughput network service with Tokio, and I'm running into an issue where the memory used by a `BytesMut` grows without bound as more and more messages are processed. After some debugging, I was able to come up with this minimal example:

    extern crate bytes;
    use bytes::*;

    fn leaks_memory() {
        let mut data = BytesMut::from(Vec::with_capacity(16));

        loop {
            data.reserve(4);
            data.put(&b"leak"[..]);
            data.advance(4);
        }
    }

    fn also_leaks_memory() {
        let mut data = BytesMut::with_capacity(8192);

        loop {
            data.reserve(4);
            data.put(&b"leak"[..]);
            data.advance(4);
        }
    }

    fn constant_memory() {
        let mut data = BytesMut::from(Vec::with_capacity(16));

        loop {
            data.reserve(4);
            data.put(&b"leak"[..]);

            data = data.split_off(4);
        }
    }

    fn also_constant_memory() {
        let mut data = BytesMut::with_capacity(16);

        loop {
            data.reserve(4);
            data.put(&b"leak"[..]);
            data.advance(4);
        }
    }

It looks like what's happening is that a `BytesMut` backed by a `Vec` isn't actually reusing space at the beginning of the buffer when `reserve` is called, but only if it was moved forwards with `advance`. I'm not sure whether any of the other methods (e.g. `truncate`) cause the problem. The documentation says `reserve` will try to recycle existing space (which it's clearly doing in `also_constant_memory`) but it seems to be failing when it shouldn't.

**Comment by carllerche** [comment]
I see what is going on. This is kind of the intended behavior as `advance` though I can see how the docs could be confusing.

That said, the case you are illustrating could be optimized by "resetting" the cursors when the read cursor and the write cursor line up.

Do you want to take a stab at implementing that?

**Comment by nzentzis** [comment]
The `advance` implementation looks entirely reasonable to me. I did implement a fix for the issue though. In `reserve`, when `kind == KIND_VEC`, instead of just calling `Vec::reserve` I check whether there's free space at the beginning of the buffer and copy everything back to the start if so.

With the fix implemented, the test cases I showed all run in constant memory. Does it seem like a reasonable approach to you? If so, I can open a PR.
